### PR TITLE
nucleus in editor, locked organelle, gui change #677 #678 #679

### PR DESCRIPTION
**Pull request by Dak2896**
- Nucleus can be add in Editor
- Can be add one Nucleus
- Some organelle to be added need nucleus
- GUI has locked and unlocked state for organelles that needs nucleus
- Nucleus has icon

### CODE REVIEW

**Comment by Dak2896** [comment]
The thing that was to change was how function that update selection on organelle work.

scripts/gui/microbe_editor.mjs line 243 from 234

"This should be in the upper function. This is could for example be called from the editor, so this must take the organelle name as a parameter."

File: scripts/gui/microbe_editor.mjs [hunk 1]
```diff
@@ -229,42 +238,42 @@ function setRedo(enabled){
 }
 
 //! Sends organelle selection to the Game
-function onSelectNewOrganelle(organelle){
+function onSelectNewOrganelle(element){
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: This still has the problem that this takes in the DOM element.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 2]
```diff
@@ -562,6 +576,25 @@ class MicrobeEditor{
         // LOG_WRITE("Mouse hex: " + qr + ", " + rr);
     }
 
+    bool checkIsNucleusPresent()
+    {
+        for(uint i = 0; i < editedMicrobe.length(); ++i){
+            auto organelle = cast<PlacedOrganelle>(editedMicrobe[i]);
+            if (organelle.organelle.name=="nucleus"){
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Some spaces around `==` would be nice

File: scripts/gui/microbe_editor.mjs [hunk 3]
```diff
@@ -229,42 +238,42 @@ function setRedo(enabled){
...
-        updateSelectedOrganelle(organelle);
+        updateSelectedOrganelle(element.organelle);
     }
 }
 
 //! Updates the GUI buttons based on selected organelle
 function updateSelectedOrganelle(organelle){
 
-    // Remove the selected text from existing ones
+    // Make all buttons unselected except the one that is now selected
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: I *think* you broke the logic here. I don't think it properly clears the properties anymore.

Comment (Dak2896) [reply scripts/gui/microbe_editor.mjs]: The only bug now i have if in the case there is nucleus i select for example mitochondria, then i undo nucleus mitochondria it's shown as selected but you can't put it, if you just select another one enabled then you can't select again mitochondria.

File: scripts/gui/microbe_editor.mjs (hunk 1 above)
Comment (Dak2896) [reply scripts/gui/microbe_editor.mjs]: i pass all the element, both DOM and organelle, element is for each variable of organelleSelectionElements

File: scripts/gui/microbe_editor.mjs (hunk 3 above)
Comment (hhyyrylainen) [reply scripts/gui/microbe_editor.mjs]: I think the previous logic was better, so you should revert the changes to this function and put the deselection checks to the actual click callback, like I said previously that I strongly believe that this method *must* be made to take in a string and not a DOM object.

File: scripts/gui/microbe_editor.mjs (hunk 1 above)
Comment (hhyyrylainen) [reply scripts/gui/microbe_editor.mjs]: I want this function to be as generic as possible. So this should only take in the name of the organelle.

File: scripts/gui/microbe_editor.mjs [hunk 4]
```diff
@@ -126,7 +130,10 @@ export function setupMicrobeEditor(){
 
         element.element.addEventListener("click", (event) => {
             event.stopPropagation();
-            onSelectNewOrganelle(element.organelle);
+
+            if(!$(element.element).hasClass("DisabledButton")) {
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: I think using jquery here is unnecessary as the modern dom classlist object should have a contains method.

Comment (Dak2896) [reply scripts/gui/microbe_editor.mjs]: Fixed

**Comment by Dak2896** [comment]
Function checks updated.

File: scripts/gui/microbe_editor.mjs (hunk 3 above)
Comment (hhyyrylainen) [reply scripts/gui/microbe_editor.mjs]: The comment on this line is incorrect after your changes. You need to adjust the comments in this function as you changed the logic.

**Comment by Dak2896** [comment]
I update one comment and Jenkins get broked, don't know why.

File: scripts/gui/microbe_editor.mjs [hunk 5]
```diff
@@ -241,26 +252,28 @@ function onSelectNewOrganelle(organelle){
     }
 }
 
-//! Updates the GUI buttons based on selected organelle
+//! Updates the GUI buttons based on selected organelle and button status
 function updateSelectedOrganelle(organelle){
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: I just realized something. Why does this function need to be changed? The user is blocked from selecting a disabled organelle in the click callback. This function should work like it was before.

Comment (Dak2896) [reply scripts/gui/microbe_editor.mjs]: that function isn't called in else branch of if(common.isInEngine) so with some test (maybe done bad) i suppose that updateOrganelle was called every iteration with GUI.

**Comment by Untrustedlife** [comment]
Ill have to check this out and do a review, for now, from glancing it seems basically ready

File: scripts/gui/thrive_gui.html [hunk 6]
```diff
@@ -398,10 +398,18 @@
...
-               </span>
+                <br>
+                Can be used to process the normally toxic soup containing hydrogen sulfide that comes out of hydrothermal vents at the bottom of the ocean. Into glucose.
+                </span>
                 <div id="ChemoplastIcon"></div>Chemoplast<br>40 MP</td>
+            <td id="addNucleus" class="OrganelleListItem">
+                <span class="tooltiptext">
+                Eukariote Nucleus<hr><br> Cost: 100 mutation points<hr><br>
+                Performs Process: Eukaryote Evolution<hr><br>
+                Allows for synthesis of RNA, MRNA, allow add chloroplasts and mitochondria
```
Comment (Untrustedlife) [inline scripts/gui/thrive_gui.html]: I dont want to hAve to commit a tooltip change after we merge so can i suggest changing this line to:

Allows for synthesis of RNA, MRNA, allows cell to evolve membrane bound organelles.


Also change the displayed name in the tool tip and the button to just "Nucleus". Not  "Eukariote Nucleus"

Comment (Untrustedlife) [reply scripts/gui/thrive_gui.html]: Also please make it scale the player cell down to 50% like the bacteria do right now. That is nessessary.

Comment (Untrustedlife) [reply scripts/gui/thrive_gui.html]: Or we can make that a seperate issue i suppose

File: scripts/gui/microbe_editor.mjs [hunk 7]
```diff
@@ -249,18 +260,19 @@ function updateSelectedOrganelle(organelle){
 
         if(element.element.contains(selectedOrganelleListItem)){
             element.element.removeChild(selectedOrganelleListItem);
+            element.element.classList.remove("Selected");
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: This method still has changes that I don't think are necessary. This logic is less foolproof

File: scripts/gui/thrive_gui.html (hunk 6 above)
Comment (Dak2896) [reply scripts/gui/thrive_gui.html]: The reduction is like is made in Microbe operations if is bacteria the species right, right?

**Comment by Dak2896** [comment]
I completely copy old functions, so we are sure is not broken.

**Comment by hhyyrylainen** [comment]
I'll test this in the next few days and then merge, unless @Untrustedlife has some further issues with this.

**Comment by Untrustedlife** [comment]
We will just make the scaling its own issue. I think

**Comment by Dak2896** [comment]
> We will just make the scaling its own issue. I think

Ok, i already make some test, they worked but still some problem with render and species system, but we wil speak better about that in next issue you right

**Comment by Dak2896** [comment]
I just change a little thing, i make also button for nucleus disabled when nucleus is already add, with undo and redo work as other organelles sensible to nucleus presence.

I'm thinking about if the check to allow or not put organelles is necessary in script code?
from GUI now it's prevent the send to the script, but of course if html code is changing with inspect you can hack the editor, with check in script  it doesn't work more.

**Comment by hhyyrylainen** [comment]
So I tested this and I immediately noticed two issues:
The editor button for nucleus still has invalid name, it is tyoped eukaryotic nucleus. It should just be "Nucleus".
When adding the nucleus to the player species the ER/Golgi is rotated incorrectly (compare how the player cells nucleus looks compared to other cells).

These need to be fixed.

**Comment by Dak2896** [comment]
I just forgot to commit name in editor, about the nucleus i think the problem is that in config was rotated 180 degrees, i try to fix it.

**Comment by Dak2896** [comment]
Rotation for organelle nucleus seems working, i modify function that i do not so good, it's possible that i make mistake, i make a lot of test but it's necessary do more of them for sure.

**Comment by Untrustedlife** [comment]
It would be nice to merge this and get a dev build out soon, i havent put out a dev build in a couple days.

**Comment by hhyyrylainen** [comment]
Okay, so I deleted my build/bin folder and now the rotation is correct. I have no clue why that was necessary, but I'll merge this now.
I have a few clarifying comments to add to a few places so I'll try to remember to make them soonish.

**Pull request merged**